src/module_3/Module3_Milestone1.md

```diff
@@ -0,0 +1,1825 @@
+y_pred = val_data['global_popularity']
+y_true = val_data['outcome']
+
+
+plot_curves(y_true, y_pred, curve_type="both", dataset_type="train")
+
+```
+
+
+    
+![png](Module3_Milestone1_files/Module3_Milestone1_30_0.png)
+    
+
+
+### 3. Training an initial model with only numerical variables
+I will be trainning a logistic regression model as the target variable is binary.
+Additionally, I start by training some initial models that do not need preprocessing, that is, with only numerical (and binary) variables. 
+
+
+```python
+train_cols = numerical_cols + binary_cols
+```
+
+#### 3.1 Ridge Regression
+
+**Important:** Before training the Ridge Regression model, it is crucial to standardize the data, ensuring each column has a mean of 0 and a standard deviation of 1. This standardization makes all features comparable, giving equal importance to each one in the model. This practice prevents variables with larger scales from disproportionately influencing the model's predictions due to their scale rather than their relevance. Additionally, standardizing the data helps in achieving faster and more stable convergence of the regression algorithm.
+
+
+```python
+# First, standardize the data
+scaler = StandardScaler()
+
+# Fit the scaler on the training data and transform it
+X_train_scaled = scaler.fit_transform(X_train[train_cols])
+
+
+# Apply the same transformation to the validation data
+X_val_scaled = scaler.transform(X_val[train_cols])
+
+```
+
+
+```python
+def plot_curves_with_hyperparameter(ax, y_true, y_pred, label, curve_type="both"):
+    if curve_type in ["precision-recall", "both"]:
+        # Compute the precision-recall curve and its AUC
+        precision, recall, _ = precision_recall_curve(y_true, y_pred)
+        pr_auc = auc(recall, precision)
+        
+        ax[0].step(recall, precision, label=f'{label} (AUC={pr_auc:.2f})')
+        ax[0].set_xlabel('Recall')
+        ax[0].set_ylabel('Precision')
+        ax[0].set_title('Precision-Recall Curve')
+
+    if curve_type in ["roc", "both"]:
+        # Compute the ROC curve and its AUC
+        fpr, tpr, _ = roc_curve(y_true, y_pred)
+        roc_auc = auc(fpr, tpr)
+        
+        ax[1].plot(fpr, tpr, label=f'{label} (AUC={roc_auc:.2f})')
+        ax[1].set_xlabel('False Positive Rate')
+        ax[1].set_ylabel('True Positive Rate')
+        ax[1].set_title('ROC Curve')
+```
+
+
+```python
+# Train the Logistic Regression model
+
+# Prepare the plot environments for both training and validation
+fig_train, axes_train = plt.subplots(1, 2, figsize=(14, 6))
+fig_train.suptitle('Training Data - Performance Curves Across Different C Values')
+fig_val, axes_val = plt.subplots(1, 2, figsize=(14, 6))
+fig_val.suptitle('Validation Data - Performance Curves Across Different C Values')
+
+
+# Define hyperparameter for the level of regularization
+cs = [1e-8, 1e-6, 1e-4, 1e-2, 1, 100, 1e4, None]
+
+
+for c in cs:
+    if c == None:
+        lr = LogisticRegression(penalty = None)
```

Comment:
Tip: take a look at scikit-learn pipelines to have standardization and modelling on one step and avoid data leakages